### PR TITLE
Implement crossref

### LINKED ISSUE #224: New meta config for list (and possible other types) inheritance from other settings
For example, you might want a default list in `vlsi.*` that you can reference in `tool.*`

### PR DESCRIPTION
**Pull request by edwardcwang**
Allows settings to reference other settings arbitrarily.

Close #224

### CODE REVIEW

**Comment by jwright6323** [comment]
It looks like there's no dynamiccrossref test, which we'll probably want. Also see notes about type restrictions on crossref.

John

**Comment by edwardcwang** [comment]
@jwright6323 Added dynamic crossref test. I was not sure what you meant by type restrictions on crossref.

**Comment by jwright6323** [comment]
Apparently I needed to click another button to get the comments to show up.

File: src/hammer_config/config_src.py [hunk 1]
```diff
@@ -138,6 +139,27 @@ def meta_append(config_dict: dict, key: str, value: Any) -> None:
     def meta_subst(config_dict: dict, key: str, value: Any) -> None:
         config_dict[key] = perform_subst(value)
 
+    def meta_crossref(config_dict: dict, key: str, value: Any) -> None:
+        """
+        Copy the contents of the referenced key for use as this key's value.
+        If the reference is a list, then apply the crossref for each element
```
[the remark this answers is not in the document]
Comment (edwardcwang) [reply src/hammer_config/config_src.py]: `value` is the way that we specify what we want to reference. e.g. if `value` is `target.tool.blah`, then `key`'s value is replaced with the value of the key `target.tool.blah`.

If `value` is an int/float/etc, it doesn't make sense to get the value of the key `True` or the key `123.45`.

File: src/hammer_config/config_src.py [hunk 2]
```diff
@@ -138,6 +139,27 @@ def meta_append(config_dict: dict, key: str, value: Any) -> None:
...
+        of the list.
+        """
+        if type(value) == str:
+            config_dict[key] = config_dict[value]
+        elif type(value) == list:
+            def check_and_get(k: Any) -> Any:
+                if not isinstance(k, str):
+                    raise ValueError("crossref (if used with lists) can only be used only with lists of strings")
```
[the remark this answers is not in the document]
Comment (edwardcwang) [reply src/hammer_config/config_src.py]: This is referring to `value` which doesn't affect what you can reference (you are free to reference strings, bools, dicts, etc). You just can't get the value of `{}`. or `[123]`.

[the remark this answers is not in the document]
Comment (jwright6323) [reply src/hammer_config/config_src.py]: What would a list of keys be used for then? Does it take the last element, merge them, or return a list of all of the `crossref`ed keys?

File: src/hammer_config/config_src.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jwright6323) [reply src/hammer_config/config_src.py]: Makes sense.

File: src/hammer_config/config_src.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (edwardcwang) [reply src/hammer_config/config_src.py]: It `crossref`s each element. See https://github.com/ucb-bar/hammer/pull/226/files#diff-f885e8454d11cd2eb8d317074b746844R252

[the remark this answers is not in the document]
Comment (jwright6323) [reply src/hammer_config/config_src.py]: OK makes sense.

**Pull request merged**